orttraining/orttraining/training_api/checkpoint_property.cc

```diff
@@ -23,43 +25,51 @@ TypedCheckpointProperty<T>::TypedCheckpointProperty(const ONNX_NAMESPACE::Tensor
   std::vector<T> data_vector(1);
   T* p = data_vector.data();
   ORT_THROW_IF_ERROR(utils::UnpackTensor<T>(tensor_proto, model_path, p, expected_num_elements));
-  prop_name_ = tensor_proto.name();
-  prop_value_ = data_vector[0];
+  name = tensor_proto.name();
+  value = data_vector[0];
 }
 
-template <typename T>
-ONNX_NAMESPACE::TensorProto TypedCheckpointProperty<T>::ToTensorProto() {
-  auto t_proto = ONNX_NAMESPACE::ToTensor<T>(prop_value_);
-  t_proto.set_name(prop_name_);
-  return t_proto;
-}
-
-namespace {
-
-std::shared_ptr<CheckpointProperty> CreateCheckpointPropertyFromTensorProto(
-    const ONNX_NAMESPACE::TensorProto& tensor_proto) {
+CheckpointProperty::CheckpointProperty(const ONNX_NAMESPACE::TensorProto& tensor_proto) {
   auto data_type = tensor_proto.data_type();
   switch (data_type) {
     case ONNX_NAMESPACE::TensorProto::FLOAT: {
-      return std::static_pointer_cast<CheckpointProperty>(
-          std::make_shared<TypedCheckpointProperty<float>>(tensor_proto));
+      ParsePropertyFromTensorProto<float>(tensor_proto, prop_name_, prop_value_);
       break;
     }
     case ONNX_NAMESPACE::TensorProto::STRING: {
-      return std::static_pointer_cast<CheckpointProperty>(
-          std::make_shared<TypedCheckpointProperty<std::string>>(tensor_proto));
+      ParsePropertyFromTensorProto<std::string>(tensor_proto, prop_name_, prop_value_);
       break;
     }
     case ONNX_NAMESPACE::TensorProto::INT64: {
-      return std::static_pointer_cast<CheckpointProperty>(
-          std::make_shared<TypedCheckpointProperty<int64_t>>(tensor_proto));
+      ParsePropertyFromTensorProto<int64_t>(tensor_proto, prop_name_, prop_value_);
       break;
     }
     default:
       ORT_THROW("Unsupported input data type of ", data_type);
   }
 }
-}  // namespace
+
+ONNX_NAMESPACE::TensorProto CheckpointProperty::ToTensorProto() const {
+  onnx::TensorProto t_proto;
+  if (std::holds_alternative<float>(prop_value_)) {
```

Comment:
checking both holds_alternative and get_if is redundant
```suggestion
  if (const float* fval = std::get_if<float>(&prop_value_); fval != nullptr) {
```
or consider using `std::visit`